File: backend/app/services/portfolio_service.py
from fastapi import HTTPException
from app.db.session import get_database

async def get_user_portfolio(user_id: str):
    db = get_database()
    portfolio = await db["portfolios"].find_one({"user_id": user_id})
    if not portfolio:
        raise HTTPException(status_code=404, detail="Portfolio not found")
        
    portfolio["_id"] = str(portfolio["_id"])
    
    # In a real app, current_price is dynamically updated via the market service. 
    # We leave that mapping for the frontend or a periodic cron.
    total_holdings_value = sum([h["quantity"] * h["current_price"] for h in portfolio.get("holdings", [])])
    
    portfolio["total_value"] = portfolio["balance"] + total_holdings_value
    portfolio["pnl"] = portfolio["total_value"] - 100000.0 # initial balance
    
    return portfolio

File: spring-boot-backend/src/main/java/com/stockpredictor/service/PortfolioService.java
package com.stockpredictor.service;

import com.stockpredictor.entity.Holding;
import com.stockpredictor.entity.Portfolio;
import com.stockpredictor.repository.PortfolioRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;

@Service
public class PortfolioService {

    @Autowired
    private PortfolioRepository portfolioRepository;

    public Map<String, Object> getUserPortfolio(Long userId) {
        Portfolio portfolio = portfolioRepository.findByUserId(userId)
                .orElseThrow(() -> new RuntimeException("Portfolio not found"));

        double totalHoldingsValue = portfolio.getHoldings().stream()
                .mapToDouble(h -> h.getQuantity() * h.getCurrentPrice())
                .sum();

        double totalValue = portfolio.getBalance() + totalHoldingsValue;
        double pnl = totalValue - 100000.0; // initial balance

        Map<String, Object> response = new HashMap<>();
        response.put("id", portfolio.getId());
        response.put("userId", portfolio.getUserId());
        response.put("balance", portfolio.getBalance());
        response.put("holdings", portfolio.getHoldings().stream().map(h -> {
            Map<String, Object> holdingMap = new HashMap<>();
            holdingMap.put("symbol", h.getSymbol());
            holdingMap.put("quantity", h.getQuantity());
            holdingMap.put("averagePrice", h.getAveragePrice());
            holdingMap.put("currentPrice", h.getCurrentPrice());
            return holdingMap;
        }).toList());
        response.put("totalValue", totalValue);
        response.put("pnl", pnl);

        return response;
    }
}
